src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -107,6 +159,11 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(FoldingRangeParams reque
 
         List<FoldingRange> mappedRanges = new();
 
+        var razorFileSyntaxWalker = new RazorFileFoldingRangeSyntaxWalker(codeDocument.Source);
+        razorFileSyntaxWalker.Visit(codeDocument.GetSyntaxTree().Root);
```

Comment:
How often does this endpoint get called? Unless the client is only going to call it when there is a change to the document, I'm a little worried that we're walking the entire syntax tree here. 99% of the time, using directives will all be direct children of the root node, and will all be at the start of the file.

If we cache the results of the walk in the `documentContext.GetCodeDocumentAsync().Items` dictionary then it should ensure we only walk the tree when the document has changed. See how this method does the same thing, to avoid a similar tree walk: https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs#L964

Reply:
The endpoint seemed to be called only when there is a document change, so I'm not sure that's necessary?